packages/@ember/-internals/glimmer/lib/components/link-to.ts

```diff
@@ -838,14 +838,14 @@ const LinkComponent = EmberComponent.extend({
         )
       );
 
-      if (DEBUG && this.query === UNDEFINED) {
-        let { _models: models } = this;
-        let lastModel = models.length > 0 && models[models.length - 1];
-
-        assert(
-          'The `(query-params)` helper can only be used when invoking the `{{link-to}}` component.',
-          !(lastModel && lastModel.isQueryParams)
-        );
+      let { _models: models } = this;
+      if (models.length > 0) {
```

Comment:
I think we need to keep this.query === UNDEFINED (and maybe assert you can’t have both)

Reply:
@chancancode Can you please help me understand exactly what you mean? What should the code be?

Reply:
```js
if (this.query === UNDEFINED) {
  let { _models: models } = this;

  if (models.length > 0) {
    let maybeQueryParams = models[models.length - 1];

    if (isQueryParams(maybeQueryParams)) {
      this.query = maybeQueryParams.values;
      models.pop();
    }
  }
} else {
  assert('Cannot pass a QueryParams object in @models and @query at the same time', this.models.length === 0 || isQueryParams(this.models[this.models.length - 1]));
}
```

```ts
// packages/@ember/-internals/routing/lib/system/query_params.ts

import { Option } from '@glimmer/interfaces';

export default class QueryParams {
  readonly values: Option<object>;
  readonly isQueryParams = true;

  constructor(readonly values: Option< object> = null) {
  }
}

export function isQueryParams(obj: unknown): obj is QueryParams {
  return typeof obj === 'object' && obj !== null && obj['isQueryParams'] === true;
}
```

Reply:
I actually don't know why the `QueryParams` constructor accepts `null` here, is that a public API?

Reply:
@chancancode - In general, I agree that you shouldn't be able to set both `@query` and pass a `(query-param` as the last model, however I don't think we can check `this.query === UNDEFINED` unless we also change where we save off the query param value. Given the snippet you suggested above, we would only update `this.query` the first time (not for each subsequent `didReceiveAttrs` where we could have a completely new value).

Reply:
@chancancode I don't understand the `query_params.ts` snippet.
This will change the behavior of the class. Why is values not a property anymore?

Reply:
Okay sorry I think I understand the problem now. I know how to fix it but I'll do it later. You can revert to what you had before here (removing the `this.query === UNDEFINED` check and the assertion I added).

Reply:
@Alonski in TypeScript, `constructor(readonly values: Option<object> = null)` is a shorthand for declaring the field and assigning it from the constructor:

```ts
export default class QueryParams {
  constructor(readonly values: Option< object> = null) {
  }

  get isQueryParams(): true {
    return true;
  }
}
```

...is the same as...

```ts
export default class QueryParams {
  readonly values: Option<object>;

  constructor(values: Option< object> = null) {
    this.values = values;
  }

  get isQueryParams(): true {
    return true;
  }
}
```